services/ui-src/src/components/modals/AddEditEntityModal.tsx

```diff
@@ -0,0 +1,109 @@
+import { useContext, useState } from "react";
+import uuid from "react-uuid";
+// components
+import { Form, Modal, ReportContext } from "components";
+import { Text } from "@chakra-ui/react";
+import { Spinner } from "@cmsgov/design-system";
+// utils
+import { AnyObject, ReportStatus } from "types";
+import { useUser } from "utils";
+
+export const AddEditEntityModal = ({
+  entityType,
+  modalData,
+  selectedEntity,
+  modalDisclosure,
+}: Props) => {
+  const { report, updateReport } = useContext(ReportContext);
+  const { full_name } = useUser().user ?? {};
+  const [submitting, setSubmitting] = useState<boolean>(false);
+
+  const { form, addTitle, editTitle, message } = modalData;
+
+  const writeEntity = async (formData: any) => {
+    setSubmitting(true);
+    const submitButton = document.querySelector("[form=" + form.id + "]");
+    submitButton?.setAttribute("disabled", "true");
+
+    const reportKeys = {
+      state: report?.state,
+      id: report?.id,
+    };
+    const currentEntities = report?.fieldData?.[entityType] || [];
+    if (selectedEntity?.id) {
+      // if existing entity selected, edit
+      const selectedEntityIndex = currentEntities.indexOf(
+        (entity: AnyObject) => entity.id === selectedEntity.id
+      );
+      const updatedEntities = currentEntities.splice(selectedEntityIndex, 1, {
+        id: selectedEntity.id,
+        ...formData,
+      });
+      await updateReport(reportKeys, {
+        lastAlteredBy: full_name,
+        reportStatus: ReportStatus.IN_PROGRESS,
+        fieldData: {
+          [entityType]: updatedEntities,
+        },
+      });
+    } else {
+      // create new entity
+      const datatowrite = {
+        lastAlteredBy: full_name,
+        reportStatus: ReportStatus.IN_PROGRESS,
+        fieldData: {
+          [entityType]: [
+            ...currentEntities,
+            {
+              id: uuid(),
+              ...formData,
+            },
+          ],
+        },
+      };
+      await updateReport(reportKeys, datatowrite);
+    }
+    setSubmitting(false);
+    modalDisclosure.onClose();
```

Comment:
nit: I like it a bit better if the dataToWrite object depends on the if/else branch, but updateReport happens regardless just above setSubmitting, outside the conditional. I like how that reads a little better, though just me

Reply:
good call. made the change.